### PR TITLE
Support hocon config and list of config file

### PR DESCRIPTION
**Pull request by jxnu-liguobin**
#585 Can't checkstyle be automatically formatted?  Idea's keymap doesn't seem to format it.

1. support HOCON configuration file.
2. support JSON configuration file.
3. mixed configuration is supported.
4. fallback is supported, that is, if the same key exists, the previous one is preferred, and the last configuration file is usually used as the default configuration.

Theoretically, all the features of config are supported such as  Java properties, JSON, and a human-friendly JSON superset.

**Note: Please note that since all configurations are merged and resolved to JSON(To deserialize to an object), even if you use the HOCON configuration, do not use its unique features, such as comments and descriptions. This will cause parsing to JSON to fail.**

Example:

HOCON conf:
```conf
generateClient = true,
generateApis = true,
generateBuilder = true,
generateImmutableModels=true,
generateToString=true,
generateEqualsAndHashCode= true
apiPackageName="io.github.graphql.j.resolver"
modelPackageName="io.github.graphql.j.model"
modelNameSuffix="TO"
apiInterfaceStrategy= "DO_NOT_GENERATE"
apiRootInterfaceStrategy= "SINGLE_INTERFACE"
generateModelsForRootTypes= true
apiNamePrefix="GitHub"
addGeneratedAnnotation=false
generatedLanguage="KOTLIN"
customTypesMapping  = {
    Long = "Long",
    Object = "org.json.JSONObject"
  }
customAnnotationsMapping = { #If you want to take it as a whole, you must add quote on it, or it will be separated.
    "QuestionNode.metaData" = ["com.fasterxml.jackson.databind.annotation.JsonDeserialize(using = com.github.dreamylost.JsonObjectDeserializer::class)"]
    "QuestionNode.envInfo" = ["com.fasterxml.jackson.databind.annotation.JsonDeserialize(using = com.github.dreamylost.JsonObjectDeserializer::class)"]
}
```

JSON conf:
```json
{
    "addGeneratedAnnotation":false,
    "apiInterfaceStrategy":"DO_NOT_GENERATE",
    "apiNamePrefix":"GitHub",
    "apiPackageName":"io.github.graphql.j.resolver",
    "apiRootInterfaceStrategy":"SINGLE_INTERFACE",
    "customAnnotationsMapping":{
        "QuestionNode.envInfo":[
            "com.fasterxml.jackson.databind.annotation.JsonDeserialize(using = com.github.dreamylost.JsonObjectDeserializer::class)"
        ],
        "QuestionNode.metaData":[
            "com.fasterxml.jackson.databind.annotation.JsonDeserialize(using = com.github.dreamylost.JsonObjectDeserializer::class)"
        ]
    },
    "customTypesMapping":{
        "Long":"Long",
        "Object":"org.json.JSONObject"
    },
    "generateApis":true,
    "generateBuilder":true,
    "generateClient":true,
    "generateEqualsAndHashCode":true,
    "generateImmutableModels":true,
    "generateModelsForRootTypes":true,
    "generateToString":true,
    "generatedLanguage":"KOTLIN",
    "modelNameSuffix":"TO",
    "modelPackageName":"io.github.graphql.j.model"
}
```

### CODE REVIEW

File: plugins/gradle/graphql-java-codegen-gradle-plugin/src/main/java/io/github/kobylynskyi/graphql/codegen/gradle/GraphQLCodegenGradleTask.java [hunk 1]
```diff
@@ -775,12 +775,12 @@ public String getResolverParentInterface() {
 
     @InputFile
```
Comment (kobylynskyi) [inline plugins/gradle/graphql-java-codegen-gradle-plugin/src/main/java/io/github/kobylynskyi/graphql/codegen/gradle/GraphQLCodegenGradleTask.java:776]: @jxnu-liguobin now it becomes @InputFiles

File: src/main/java/com/kobylynskyi/graphql/codegen/supplier/MergeableMappingConfigSupplier.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+     * @return The string of the configuration after merging.
+     */
+    private static String parseConfigAndMerged(List<String> confFiles) {
+        try {
+            if (confFiles == null || confFiles.isEmpty()) {
+                return null;
+            }
+            Optional<Config> config = confFiles.stream()
+                    .map(c -> ConfigFactory.parseFile(new File(c))).reduce(Config::withFallback);
+            ConfigRenderOptions configRenderOptions = ConfigRenderOptions.concise();
```
Comment (kobylynskyi) [inline src/main/java/com/kobylynskyi/graphql/codegen/supplier/MergeableMappingConfigSupplier.java:59]: 1. configRenderOptions can be static.
2. Returnable object can be converted to a chained method call (more functional style):
 confFiles.stream()
                  .map(c -> ConfigFactory.parseFile(new File(c)))
                  .reduce(Config::withFallback)
                  .map(value -> value.root().render(configRenderOptions))
                  .orElse(null)

What do you think?

**Review by kobylynskyi — CHANGES_REQUESTED** [review]
@jxnu-liguobin looks good to me. Good job. I just left a few minor comments (code-style-wise)

File: src/main/java/com/kobylynskyi/graphql/codegen/supplier/MergeableMappingConfigSupplier.java (hunk 2 above)
Comment (jxnu-liguobin) [reply src/main/java/com/kobylynskyi/graphql/codegen/supplier/MergeableMappingConfigSupplier.java:59]: Ok.it looks good

File: docs/codegen-options.md [hunk 3]
```diff
@@ -6,7 +6,7 @@
...
 | `outputDir`                                           | String                                                                | None                                                   | The output target directory into which code will be generated. |
-| `jsonConfigurationFile`                               | String                                                                | Empty                                                  | Path to an external mapping configuration. |
+| `configurationFiles`                                   | List(String)                                                          | Empty                                                  | Path to an external mapping configuration. It support either JSON or HOCON. The same key is used in order, so the default configuration should be placed at the end.|
```
Comment (kobylynskyi) [inline docs/codegen-options.md:9]: Paths to the files with mapping configurations. Supported formats. JSON, HOCON. Order of specified configuration files matters, so the default configuration should be placed at the end.

**Review by kobylynskyi — APPROVED** [review]
Looks good. One very minor request for a change.
Also please note that since this PR has breaking changes, then I am going to include them into release 5.0.0. So I am not going to merge this PR until we release 4.1.6.
Thanks

**Comment by jxnu-liguobin** [comment]
> Also please note that since this PR has breaking changes, then I am going to include them into release 5.0.0. So I am not going to merge this PR until we release 4.1.6.

Yes, and will there be 4.2.x?

**Review by kobylynskyi — APPROVED** [review]

**Pull request merged**